File: php/incorex.php
<?php
function api_query($api_name, array $req = array())
{
    $mt = explode(' ', microtime());
    $NONCE = $mt[1] . substr($mt[0], 2, 6);

    // API settings
    $key = "YOUR_API_KEY";       //TODO replace with your api key from profile page
    $secret = "YOUR_API_SECRET"; //TODO replace with your api secret from profile page

    $url = "https://api.incorex.com/v1/$api_name";

    $req['nonce'] = $NONCE;

    // generate the POST data string
    $post_data = http_build_query($req, '', '&');

    $sign = hash_hmac('sha512', $post_data, $secret);

    // generate the extra headers
    $headers = array(
        'Sign: ' . $sign,
        'Key: ' . $key,
    );

    // our curl handle (initialize if required)
    static $ch = null;
    if (is_null($ch)) {
        $ch = curl_init();
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
        curl_setopt($ch, CURLOPT_USERAGENT, 'Mozilla/4.0 (compatible; PHP client; ' . php_uname('s') . '; PHP/' . phpversion() . ')');
    }
    curl_setopt($ch, CURLOPT_URL, $url);
    curl_setopt($ch, CURLOPT_POSTFIELDS, $post_data);
    curl_setopt($ch, CURLOPT_HTTPHEADER, $headers);
    curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, FALSE);

    // run the query
    $res = curl_exec($ch);
    if ($res === false) throw new Exception('Could not get reply: ' . curl_error($ch));
   
    $dec = json_decode($res, true);
    if ($dec === null)
        throw new Exception('Invalid data received, please make sure connection is working and requested API exists');

    return $dec;
}

//Example
print_r(api_query("user_info"));
echo "<br>";
print_r(api_query("user_trades", array("pair" => "BTC_USD", "limit" => 100, "offset" => 0))); 

// $result = api_query("trades", array("pair" => "BTC_USD"));
// $result = api_query("order_book", array("pair" => "BTC_USD", "limit" => 100));
// $result = api_query("ticker", array());
// $result = api_query("pair_settings", array());
// $result = api_query("currency", array());

// $result = api_query("user_info", array());
// $result = api_query("order_create", array("pair" => "BTC_USD", "quantity" => 2, "price" => 100, "type" => "sell"));
// $result = api_query("order_cancel", array("order_id" => 123456));
// $result = api_query("user_open_orders", array());
// $result = api_query("user_trades", array("pair" => "BTC_USD", "limit" => 100, "offset" => 0));
// $result = api_query("user_cancelled_orders", array());
// $result = api_query("order_trades", array("order_id" => 123456));
// $result = api_query("required_amount", array("pair" => "BTC_USD", "quantity" => 2));
// $result = api_query("deposit_address", array());
// $result = api_query("withdraw_get_txid", array("task_id" => 123456));
// $result = api_query("create_xvoucher", array("currency" => "USD", "amount" => 100.00));
// $result = api_query("activate_xvoucher", array("code" => "123456"));
// $result = api_query("wallet_history", array("date" => 1525122000));
?>

File: ruby/incorex.rb
require 'net/https'
require "json"

module IncoreX
  class Error < StandardError
    attr_reader :object

    def initialize(object)
      @object = object
    end
  end

  class API
    class << self
      KEY = "YOUR_API_KEY"         # TODO replace with your api key from settings page
      SECRET = "YOUR_API_SECRET"   # TODO replace with your api secret from settings page

      def api_query(method, params = nil)
        raise ArgumentError unless method.is_a?(String) || method.is_a?(Symbol)

        params = {} if params.nil?
        params['nonce'] = nonce

        uri = URI.parse(['https://api.incorex.com/v1', method].join('/'))

        post_data = URI.encode_www_form(params)

        digest = OpenSSL::Digest.new('sha512')
        sign = OpenSSL::HMAC.hexdigest(digest, SECRET, post_data)

        headers = {
          'Sign' => sign,
          'Key'  => KEY
        }

        req = Net::HTTP::Post.new(uri.path, headers)
        req.body = post_data
        http = Net::HTTP.new(uri.host, uri.port)
        http.use_ssl = true if uri.scheme == 'https'
        response = http.request(req)

        unless response.code == '200'
          raise IncoreX::Error.new(__method__), ['http error:', response.code].join(' ')
        end

        result = response.body.to_s

        unless result.is_a?(String) && valid_json?(result)
          raise IncoreX::Error.new(__method__), "Invalid json"
        end

        JSON.load result
      end

      private

      def valid_json?(json)
        JSON.parse(json)
        true
      rescue
        false
      end

      def nonce
        Time.now.strftime("%s%6N")
      end
    end
  end
end



puts "%s" % IncoreX::API.api_query('user_info').inspect
puts "%s" % IncoreX::API.api_query('user_trades', pair: 'BTC_USD', limit: 100, offset: 0).inspect

# puts "%s" % IncoreX::API.api_query('trades', pair: 'BTC_USD').inspect
# puts "%s" % IncoreX::API.api_query('order_book', pair: 'BTC_USD', limit: 100).inspect
# puts "%s" % IncoreX::API.api_query('ticker').inspect
# puts "%s" % IncoreX::API.api_query('pair_settings').inspect
# puts "%s" % IncoreX::API.api_query('currency').inspect

# puts "%s" % IncoreX::API.api_query('user_info').inspect
# order_params = {
#      pair: 'BTC_USD',
#      quantity: 2,
#      price: 100,
#      type: 'sell'
#    }
# puts "%s" % IncoreX::API.api_query('order_create', order_params).inspect
# puts "%s" % IncoreX::API.api_query('order_cancel', order_id: 12345).inspect
# puts "%s" % IncoreX::API.api_query('user_open_orders').inspect
# puts "%s" % IncoreX::API.api_query('user_trades', pair: 'BTC_USD', limit: 100, offset: 0).inspect
# puts "%s" % IncoreX::API.api_query('user_cancelled_orders', limit: 100, offset: 0).inspect
# puts "%s" % IncoreX::API.api_query('order_trades', order_id: 12345).inspect
# puts "%s" % IncoreX::API.api_query('required_amount', pair: 'BTC_USD', quantity: 2).inspect
# puts "%s" % IncoreX::API.api_query('deposit_address').inspect
# puts "%s" % IncoreX::API.api_query('withdraw_get_txid', task_id: 123456).inspect
# puts "%s" % IncoreX::API.api_query('create_xvoucher', currency: 'USD', amount: 100.00).inspect
# puts "%s" % IncoreX::API.api_query('activate_xvoucher', code: '123456').inspect
# puts "%s" % IncoreX::API.api_query('wallet_history', date: 1525122000).inspect
